test/diff/expected-stdout

```diff
@@ -4,7 +4,7 @@
 
 Diff:
 * B bar | changed: description
-* C bar/extractor/ex-generic-v2/empty | changed: name
+* C bar/extractor/ex-generic-v2/empty | changed: metadata, name
```

Comment:
Stačí tento func test, nebo to mám přidat i do nějakých dalších?

Reply:
Potrebujes este test na `diff-details`, aby sme videli, ako vyzeraju podrobnosti.
+ ak nie su, tak aj `push/pull` testy na metadata (to kludne v dalsom PR).
Budes musel najprv vyriesit, aby `initial-state.json` podporoval nastavovanie metadat,

Reply:
um, ty podrobnosti asi nemají vypadat jako json, co? 
![CleanShot 2022-01-24 at 11 45 17](https://user-images.githubusercontent.com/215660/150768982-07c71619-7732-4741-a4d7-51a3901e7aac.jpg)
Koukám do `diff` package ale není mi teda moc jasný jak s tím něco udělat 🙂

Reply:
... ono, keby tam nebola nikdy ta `nil` hodnota, ale vzdy prazdna map, tak myslim, ze by to bolo automaticky spravne.

Takze bud upravit metody/kod, aby tam nebola `nil` (ked prejdu vsetky mappers, tj. remote load, local load).
Alebo sa to fixne az v `diff` balicku.

Da sa to jednoducho upravit, aby tam nebolo `nil`? Alebo mam pozriet, ako to treba upravit v `diff`?
https://github.com/keboola/keboola-as-code/blob/4a32171979f165ade569be3c4bfc74f05fe15193/internal/pkg/diff/diff.go#L183-L203

Reply:
Aha, jo to by myslím šlo, tam dát prázdnou map. Zkusím. 👍

Reply:
Nó, tak nevím, asi jsem se do toho zamotal. 😅 Při local load dám do `config.Metadata` prázdnou map, při local save do `manifest.Metadata` prázdnou order map. Zkusil jsem i odstranit `omitempty` z json tagu `Config.Metadata`: https://github.com/keboola/keboola-as-code/pull/470/commits/8ed241b2ab456f2e18c8287b9e017d2c0b4318ef ale ten `nil` mi tam pořád zůstává.

Reply:
pozriem za chvilku

Reply:
@JakubMatejka 

Diff porovnava `remote` a `local` object.
Local vyzera dobre.
Ale v `remote.go` mas:
```
if !found {
	return
}
```
A myslim, ze `nil` sa berie tam (teda nenastavuje sa nic namiesto `nil`).

Reply:
Malo by tam byt asi:
```
if !found {
	metadata := make(map[string]string)
}
```

Reply:
Aha, rozumím. Akorát tomuhle nerozumím 🙈😅
![CleanShot 2022-01-24 at 16 03 25](https://user-images.githubusercontent.com/215660/150808033-637b56d3-e858-45de-8d0e-d837dae53f55.jpg)

Reply:
@JakubMatejka neznamena to, ze je to vo vystupe nizsie ... nie od riadku `33` ale od riadku `40`?
Ked sa teda na to pytas .... alebo mi zavolaj.

Reply:
Jo to je pravda, mate mě tam ten řádek 29.
Nicméně prohození pořadí `name` a `metadata` nepomohlo:
![CleanShot 2022-01-24 at 16 09 39](https://user-images.githubusercontent.com/215660/150810026-042764d6-ab72-452e-8dc9-cf276dfd52da.jpg)
A vrácení zpátky najednou pomohlo 😃
![CleanShot 2022-01-24 at 16 15 51](https://user-images.githubusercontent.com/215660/150810271-6692126d-93f7-4369-b658-3b8c88c40a7e.jpg)
Což teda moc nechápu, ale nebudu do toho vrtat. 😅

Reply:
`%w` zodpoveda tym 4x space, takze to je OK, ... len v tom diffe nie je naprogramovane nahradzovanie whitespaces.
Je to len tak pomocka, ako najst chybu. Vyssie nad tym je cely text aj regexp, ktory to kontroluje.

Ked to uz ide tak mozno nejaky biely znak, alebo to neulozilo? neviem :)

Reply:
Stalo se mi to dneska už podruhý, že si stěžoval na to pořadí, výměna nepomohla a po vrácení zpět to fungovalo. Zpracovává diff ty json klíče vždycky ve stejným pořadí? (Podle abecedy zřejmě ne když teď čeká `name` před `metadata`.) 

Nicméně dívám se teď teda že tahle úprava rozstřelila hromadu jiných testů, hádám že všude kde se vytváří `model.Config`, tak bude třeba initnout v `Metadata` prázdnou mapu ručně. Nebylo by teda nakonec lepší upravit ten diff jak jsi navrhoval na začátku? 😄  Nebo mám pofixovat ty testy? (Mně je to jedno, jen abych třeba zbytečně nebordelil ten kód pokud to jde udělat elegantněji.)

![CleanShot 2022-01-24 at 16 28 35](https://user-images.githubusercontent.com/215660/150812829-6b0052fa-29dc-499d-b613-536fdfeda39c.jpg)

Reply:
1. moze tam byt bug, pozriem to zajtra
2. aha :) uz som si spomenul, preco sme tam ten `nil` chceli nechat.
- ako pozeram, tak to spadlo vacsinou vo funkcionalnych testoch ... tak tam by si ten nil mohol nechat ... v tej casti kodu, co kopiruje `object.metadata -> manifest.metadata` (ked nepomoze `omitempty`), ... tj. ak nie su ziadne zaznamy, tak v manifest bude `nil`

nechajme to na zajtra, pozrieme kolko by to bolo prace fixnut tie testy, a ktora varianta je lepsia.

Reply:
ok, díky 👍

Reply:
1. musime upravit `result.ChangedFields` aby zachovavali poradie:
https://github.com/keboola/keboola-as-code/blob/2976d4da18c667e944b554e575bb61e827c1ddf7/internal/pkg/diff/diff.go#L157-L160

Reply:
2. 

`configMetadataMapper.MapBeforeLocalSave`:

```
	manifest.Metadata = config.MetadataOrderedMap()
	if len(manifest.Metadata.Keys()) == 0 {
		manifest.Metadata = nil
	}
```

Reply:
jj